### PR TITLE
Rename channel to fileReader/fileWriter

### LINKED ISSUE #18112: I/O module: Should we rename I/O channels to reader and writer?
Today we have a `channel` type for use in the I/O system. An I/O channel can either be a reading channel or a writing channel.

However, as part of GSoC we are looking at adding Go-style channels to Chapel (see #17954). The immediate problem is - what should these be called, since `channel` is already used in I/O?

This issue proposes that we deprecate the term `channel` for this and instead have types `reader` and `writer`. Then we can add a generic type that matches either of these (for example, `readerOrWriter` `stream`, `iostream`, or `iochannel`) that would be supported by an interface eventually but might be directly supported in the compiler for now.

E.g. in this typical use case from the fileIO primer:
```
    // Get a binary writing channel for the start of the file.
    var w = f.writer(kind=ionative);
```

it would be reasonable for the type of `w` to be `writer` rather than `channel(writing=true)`.

| Language | I/O reader equivalent | I/O writer equivalent | I/O reader or writer |
| -- | --- | --- | --- |
| C++ | istream | ostream | ios |
| Java | Reader / BufferedReader | Writer / BufferedWriter | - |
| Rust | Read / BufRead / BufReader | Write / BufWriter | - |
| Python | BufferdReader / BufferedRandom | BufferedWriter / BufferedRandom | - |
| Go | Reader / File | Writer / File | - |

C, Swift have File but not a stream idea. Julia has a similar design but seems to call the file type IOStream.

### PR DESCRIPTION
**Pull request by mppf**
Resolves #18112

This PR renames the IO module type `channel` to two types: `fileReader` and `fileWriter`. It deprecates type `channel`.

In order to minimize the changes to the IO module, this PR uses the approach of renaming `record channel` to `record _channel` and then using a deprecated type alias `type channel = _channel` for the deprecated case and a partial instantiation `type fileReader = _channel(writing=false)` for the new types.

In order for the above to work, it was necessary to fix a compiler bug in handling of partial instantiations.

In order to get good error messages even with the `_channel` workaround, this PR adjusts the compiler to print out `fileReader` / `fileWriter` / `channel` in error messages. It makes a similar adjustment to `chpldoc` in order to fix the docs build.

This PR also renames the no-doc'd field `home` in file and channel to `_home` to match the convention of implementation-private fields beginning with `_` as with `_channel_internal`.

This PR also updates two Curl tests that don't run nightly to fix a problem with a missing `use` statement.

Reviewed by @lydia-duncan - thanks!

- [x] full local testing

### CODE REVIEW

**Comment by riftEmber** [comment]
It's unclear to me how this is connected to https://github.com/chapel-lang/chapel/issues/20501, could you clarify?

**Comment by mppf** [comment]
Oops, I had the wrong issue number. Thanks for pointing it out. I've edited it to fix.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -5482,19 +5571,19 @@ record itemReaderInternal {
 // And now, the toplevel items.
 
 /* standard input, otherwise known as file descriptor 0 */
-const stdin:channel(false, iokind.dynamic, true);
+const stdin:_channel(false, iokind.dynamic, true);
```
Comment (lydia-duncan) [inline modules/standard/IO.chpl:5574]: Would it make sense to turn this into a fileReader officially?  And the other two into fileWriters?

File: modules/standard/List.chpl [hunk 2]
```diff
@@ -1723,7 +1723,7 @@ module List {
 
       :arg ch: A channel to write to.
     */
-    proc writeThis(ch: channel) throws {
+    proc writeThis(ch: _channel) throws {
```
Comment (lydia-duncan) [inline modules/standard/List.chpl:1726]: This should be a fileWriter, right?

File: test/deprecated/iochannel.chpl [hunk 3]
```diff
@@ -0,0 +1,13 @@
+use IO;
```
Comment (lydia-duncan) [inline test/deprecated/iochannel.chpl:1]: There's an IO sub-directory in test/deprecated, so it might make sense to move it there (but I'd totally understand you being tired of this change so it's okay to not move it since it will hopefully get removed soon)

File: test/functions/generic/genericChannelInInit1.good [hunk 4]
```diff
@@ -1,5 +1,3 @@
 genericChannelInInit1.chpl:9: In initializer:
-genericChannelInInit1.chpl:9: error: cannot default-initialize a variable with generic type
-genericChannelInInit1.chpl:9: note: 'this' has generic type 'borrowed C'
-genericChannelInInit1.chpl:9: note: cannot find initialization point to split-init this variable
-genericChannelInInit1.chpl:11: note: 'this' is used here before it is initialized
+genericChannelInInit1.chpl:9: error: Cannot default-initialize a variable with generic type
```
Comment (lydia-duncan) [inline test/functions/generic/genericChannelInInit1.good:2]: Nice, this seems like a great improvement!

**Review by lydia-duncan — COMMENTED** [review]
It looks like you ended up being a bit inconsistent on when you switch to `_channel` and when you switch to `fileReader`/`fileWriter`.  I haven't called out every spot where I think it should be the latter instead because it ended up being quite a few places (but I think on a file by file basis rather than too mixed within one - it looks like Set, Map, List and a bunch of stuff in IO got updated to `_channel`).  I think it would make sense to update these now rather than waiting until we remove _channel

File: modules/standard/IO.chpl [hunk 5]
```diff
@@ -2792,7 +2881,7 @@ proc openreader(path:string,
                 start:int(64) = 0, end:int(64) = max(int(64)),
                 hints:iohints,
                 style:iostyle)
-    : channel(false, kind, locking) throws {
+    : _channel(false, kind, locking) throws {
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl:2884]: I don't know of an issue. I'll try updating this one.

File: modules/standard/IO.chpl [hunk 6]
```diff
@@ -2150,50 +2131,158 @@ record channel {
...
+cannot be assigned within the function.
+
+The default value of the :record:`fileReader` type is not associated
+with any file, and so cannot be used to perform I/O.
+
+The :record:`fileReader` type is generic.
+
+The :record:`fileReader` type supports 3 fields:
+
+ * ``param writing: bool = false``:
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl:2155]: It's my intention that it always will be `false` in order to support generic programming. However, it's something we could/should discuss in an issue. (Not immediately sure if we would need a new issue or an existing issue would cover it).

File: compiler/resolution/functionResolution.cpp [hunk 7]
```diff
@@ -7692,6 +7692,12 @@ static void resolveInitField(CallExpr* call) {
         Type* exprType = fs->defPoint->exprType->typeInfo();
         if (exprType == dtUnknown) {
           fs->type = srcType;
+        } else if (exprType->symbol->hasFlag(FLAG_GENERIC)) {
+          // e.g. the type so far is a partial instantiation
+          // but the field declaration is concrete
+          // TODO: does this need to check srcType is compatible
+          // with any partial instantiation?
+          fs->type = srcType;
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/resolution/functionResolution.cpp:7700]: I've added a test invalid-set-param-again-type-ctor.chpl for this. It currently does not compile, but the error message definitely has room for improvement.

File: modules/packages/Curl.chpl [hunk 8]
```diff
@@ -168,10 +168,10 @@ module Curl {
   /* Returns the ``CURL`` handle connected to a channel opened with
      :proc:`URL.openUrlReader` or :proc:`URL.openUrlWriter`.
    */
-  proc getCurlHandle(ch:channel):c_ptr(CURL) throws {
+  proc getCurlHandle(ch:_channel):c_ptr(CURL) throws {
```
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply modules/packages/Curl.chpl:171]: My main trouble is that the documentation will now refer to a type that cannot be found in it

[the remark this answers is not in the document]
Comment (mppf) [reply modules/packages/Curl.chpl:171]: I don't care to argue about it further so I just added the where clause :)

File: modules/standard/IO.chpl [hunk 9]
```diff
@@ -3868,7 +3958,7 @@ proc channel.writeIt(const x) throws {
      :throws SystemError: When an IO error has occurred.
    */
   deprecated "the <~> operator is deprecated"
-  inline operator channel.<~>(const ref ch: channel, const x) const ref throws
+  inline operator _channel.<~>(const ref ch: _channel, const x) const ref throws
```
[the remark this answers is not in the document]
Comment (mppf) [reply modules/standard/IO.chpl:3961]: It's moot. This code is no longer present because `<~>` has been removed entirely on `main`.

**Comment by mppf** [comment]
@lydia-duncan - this is ready for another look when you get a chance, thanks.

File: modules/packages/Curl.chpl [hunk 10]
```diff
@@ -168,10 +168,11 @@ module Curl {
   /* Returns the ``CURL`` handle connected to a channel opened with
```
Comment (lydia-duncan) [inline modules/packages/Curl.chpl:168]: It looks like there's a doc link to `IO.channel` at the top of this file (line 49) - I don't know that we'll come back to it and I don't think we get warned about broken links, so it might be good to update it now

File: modules/packages/Curl.chpl [hunk 11]
```diff
@@ -286,7 +288,8 @@ module Curl {
      :arg args: any number of tuples of the form (curl_option, value).
```
Comment (lydia-duncan) [inline modules/packages/Curl.chpl:288]: Not your fault, but it looks like the documentation for this method doesn't talk about the `ch` arg.  Might be good to add a note about it (though I'd understand if that felt like scope creep, so it'd be okay to not)

File: modules/standard/IO.chpl [hunk 12]
```diff
@@ -2134,50 +2115,158 @@ record channel {
...
+
+ *  ``param kind:iokind``:
+     kind is an enum :type:`iokind` that allows narrowing
+     this channel's I/O style for more efficient binary I/O.
+
+  * ``param locking:bool``:
+     locking is a boolean indicating whether it is safe to use this
+     channel concurrently (when `true`).
+*/
+type fileReader = _channel(writing=false, ?);
```
Comment (lydia-duncan) [inline modules/standard/IO.chpl:2149]: In looking through the documentation, I noticed that this is one of the few places `_channel` turns up.  Would you be up for using split init to hide it?

File: modules/standard/IO.chpl [hunk 13]
```diff
@@ -2134,50 +2115,158 @@ record channel {
...
+  this._readWriteThisFromLocale = x._readWriteThisFromLocale;
   this.complete();
-  on x.home {
+  on x._home {
     qio_channel_retain(x._channel_internal);
   }
 }
 
-proc channel.init=(x: channel) {
+proc _channel.init=(x: _channel) {
```
Comment (lydia-duncan) [inline modules/standard/IO.chpl:2269]: I think this needs a `pragma "no doc"`, otherwise it turns up in the documentation (and exposes the `_channel` type as a result)

**Review by lydia-duncan — APPROVED** [review]
Some documentation nits and one update-to-the-PR-message nit, but otherwise I think this is good to go!

File: test/library/packages/Curl/download-http-upload-ftp.chpl [hunk 14]
```diff
@@ -1,4 +1,5 @@
 use RunServer;
+use FileSystem;
```
[the remark this answers is not in the document]
Comment (mppf) [reply test/library/packages/Curl/download-http-upload-ftp.chpl:2]: Yes, intentional. These tests don't run everywhere & need the `use` to work. I'll mention in PR message.

**Comment by mppf** [comment]
I tried to use split init for `type channel = _channel;` to hide `_channel` as well, but that leads to `$CHPL_HOME/modules/standard/IO.chpl:2227: warning: channel type is deprecated - use fileReader or fileWriter instead`. I will just leave this `_channel` in for now.

**Comment by lydia-duncan** [comment]
Thanks for checking!
